### ISSUE TITLE
Type "property" cannot be assigned to type ""

### ISSUE THREAD

**Issue by pseudo-rnd-thoughts** [microsoft/pyright#3646]
Im one of the maintainers of [OpenAI Gym](https://github.com/openai/gym) and we are attempting to type hint the whole project and in particular strict type the core API sections (https://github.com/openai/gym/pull/2943).
However, I have been having an issue with trying to type-hint one of our subclasses. This is a minimal reproducible example for me
```python
from typing import Optional, Dict, Any

class Env:
    spec: Optional[Dict[str, Any]] = None
    render_mode: Optional[str] = None 

class Wrapper(Env):
    def __init__(self, env: Env):
        self.env = env

    @property
    def spec(self) -> Optional[Dict[str, Any]]:
        """Returns the environment specification."""
        return self.env.spec

    @property
    def render_mode(self) -> Optional[str]:
        """Returns the environment render_mode."""
        return self.env.render_mode
```
This raises this error for the `spec` function and is similar for `render_mode`. 
```
"spec" overrides symbol of same name in class "Env"
  Type "property" cannot be assigned to type "Dict[str, Any] | None"
    "property" is incompatible with "Dict[str, Any]"
    Type cannot be assigned to type "None" Pylance reportIncompatibleVariableOverride
```
This is both from terminal (`pyright example.py`) and vscode with a strict `pyproject.toml` configure in the background  

The weird part is that I have other examples where pyright doesn't complain about it
Any help would be useful to understand what pyright is thinking

`reveal_type(Wrapper.spec)` says that it is a "property"

System - VSCode `v2022.6.30`, pip Pyright `1.1.256`, python 3.10 using conda

**Comment by erictraut** [comment]
First off, I'm very excited to hear that you're adding type information to OpenAI Gym! My team uses this package, and we've had to cobble together type stubs to use it in our strictly-typed Python code base. So I applaud your efforts here!

Now, for your question...

You're attempting to override an attribute with a property. In general, this isn't safe to do. You will get different results, for example, if you access `Env.spec` (which is of type `Dict[str, Any] | None`) or `Wrapper.spec` (which is of type `property`). Also, the property in `Wrapper` is read-only, whereas the attribute in `Env` is readable, writable and deletable.

The strict-mode diagnostic check `reportINcompatibleVariableOverride` is very opinionated about this. It doesn't allow you to intermix properties and attributes. The one exception to this rule is with protocol base classes, where some limited form is allowed as long as the writability matches, but in this case pyright enforces that the property is never accessed from the class (as in `Wrapper.spec`), so it is type safe.

You have the following options:
1. Use Properties consistently in both the base class and the subclass. Either use properties in both or attributes in both.
2. Disable the `reportIncompatibleVariableOverride` diagnostic check for this file by adding a `# pyright: reportIncompatibleVariableOverride=false` to the top of the file.
3. Disable the diagnostics on a specific line by using `# type: ignore` or `# pyright: ignore[reportIncompatibleVariableOverride]`.

**Comment by pseudo-rnd-thoughts** [comment]
@erictraut Thanks, it is exciting and though painful to add type hints for a project that never intended to have them like the issue we have here.
That makes sense by `pyright` is complaining but none of the options are perfect. 
I will see makes the most sense
